### ISSUE TITLE
incorrect type inference after isinstance

### ISSUE THREAD

**Issue by sam-s** [pylint-dev/pylint#10423]
### Bug description

```python
`pylint` seems to expect that the type of the 1st `return` statement is applicable to all subsequent returns:

"https://github.com/pylint-dev/pylint/issues/"
import matplotlib.pyplot as plt
def figrow(figsize, ncols):
    "Return figure and axes for a row with ncol columns."
    if not isinstance(figsize[0], int):
        return figsize
    return plt.subplots(figsize=(ncols*figsize[0], figsize[1]),nrows=1,ncols=ncols)
fig, axes = figrow(figsize=(4,7), ncols=3)
axes[0].plot([1,2,3],[4,5,6])
fig.tight_layout()
```

### Configuration

```ini

```

### Command used

```shell
pylint pylint.py
```

### Pylint output

```python
************* Module pylint-
pylint-.py:9:0: E1136: Value 'axes' is unsubscriptable (unsubscriptable-object)
pylint-.py:10:0: E1101: Instance of 'int' has no 'tight_layout' member (no-member)
```

### Expected behavior

no diagnostics

### Pylint version

```shell
pylint 3.3.7
astroid 3.3.10
Python 3.13.2 (tags/v3.13.2:4f8bb39, Feb  4 2025, 15:23:48) [MSC v.1942 64 bit (AMD64)]
```

### OS / Environment

windows

### Additional dependencies

```python

```

**Comment by sam-s** [comment]
type annotation like
```python
def figrow(figsize, ncols) -> Tuple[Figure, List[Axes]]:
    ...
```
does not help

**Comment by sam-s** [comment]
workaround:
```python
def figrow(figsize, ncols):
    "Return figure and axes for a row with ncol columns."
    if isinstance(figsize[0], int):
        return plt.subplots(figsize=(ncols*figsize[0], figsize[1]),nrows=1,ncols=ncols)
    return figsize
```
Apparently, `pylint` infers the return type from the 1st `return` statement.

**Comment by jacobtylerwalls** [comment]
Thanks, we have a major class of bugs in pylint where by habit we default to `safe_infer()` instead of `infer_all()`.

cc/ @Pierre-Sassoulas @mbyrnepr2 @zenlyj for awareness during triage.

**Comment by jacobtylerwalls** [comment]
I mean, should this be configurable with `--strict` mode?

**Comment by Pierre-Sassoulas** [comment]
I don't think we have a strict mode in pylint. (Maybe related to adding configuration template -- discussed everywhere and in no issue in particular after 5/10mn of searching for it)
